### PR TITLE
Clarification of transport switching rules and improvement of UX

### PR DESCRIPTION
**Pull request by Akihiro-Miyazaki**
## Introduction
This proposal is to improve transport switching on the same device.

### CODE REVIEW

**Comment by Akihiro-Miyazaki** [comment]
Hi. @theresalech -san
Could you please review my proposal?
Thank you!

File: proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+* Author: [Akihiro Miyazaki (Nexty)](https://github.com/Akihiro-Miyazaki)
+* Status: **Awaiting review**
+* Impacted Platforms: [ Core / HMI ]
+
+## Introduction
+This proposal is to improve transport switching on the same device.
+
+
+## Motivation
+The rules for transport switching, such as connection priority, when the same device is connected to the Head Unit (HU) with multiple transport methods are not explicitly defined in the current SDL specifications. Therefore, designing the HU transport switching process takes time, thus inefficient.
```
Comment (theresalech) [inline proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md:13]:
```suggestion
The rules for transport switching, such as connection priority, when the same device is connected to the Head Unit (HU) with multiple transport methods are not explicitly defined in the current SDL specifications. Therefore, designing the HU transport switching process takes time, and thus is inefficient.
```

File: proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+Additionally, the user cannot know that the connection is being switched because the switching status is not notified to HMI during transport switching. Also, the user cannot operate the app screen on the HU. As a result, it causes an extremely inconvenient state for the user.
+
+
+## Proposed solution
+The following items are proposed to improve the problems mentioned above.
+1. Clarification of transport switching rules
+2. Addition of transport switching notification to HMI
+
+#### 1. Clarification of transport switching rules
+The following transport switching rules and tables should be added to [the SDLC Guidelines](https://smartdevicelink.com/en/guides/core/developer-documentation/transport-manager/) as a specification for multiple transport methods for the same device.
```
Comment (theresalech) [inline proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md:23]:
```suggestion
The following transport switching rules and tables should be added to [the SDL Core Guides](https://smartdevicelink.com/en/guides/core/developer-documentation/transport-manager/) as a specification for multiple transport methods for the same device.
```

File: proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+- The priority of transport methods: USD > Bluetooth (BT)<br>Note that WiFi cannot communicate alone by itself, therefore WiFi is excepted from the priority.
```
Comment (theresalech) [inline proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md:25]:
```suggestion
- The priority of transport methods: USB > Bluetooth (BT)<br>Note that WiFi cannot communicate alone by itself, therefore WiFi is excepted from the priority.
```

File: proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+- If the SDL connects multiple terminals (mobiles), then the device that is already connected to the SDL is prioritized.
```
Comment (theresalech) [inline proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md:26]:
```suggestion
- If the HU connects to multiple terminals (mobile devices), then the device that is already connected to the HU is prioritized.
```

File: proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+<b>Table 1.</b> Switching of Multiple Transport
+
+![Table_1_Switching_of_Multiple_Transport.png](../assets/proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX/Table_1_Switching_of_Multiple_Transport.png)
+
+*1 : If the HU recognize that current BT and new USB are connected to the same device, SDL transport will switch to USB.<br>
```
Comment (theresalech) [inline proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md:32]:
```suggestion
*1 : If the HU recognizes that current BT and new USB are connected to the same device, SDL transport will switch to USB.<br>
```

File: proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+*2 : The SDL app that uses VPM detects WiFi transport when it is launched. If the WiFi transport, which is connected to same device as the BT connection, is found, the SDL App will start VPM via WiFi.<br>
```
Comment (theresalech) [inline proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md:33]:
```suggestion
*2 : The SDL app that uses Video Projection Mode (VPM) detects WiFi transport when it is launched. If the WiFi transport, which is connected to same device as the BT connection, is found, the SDL app will start VPM via WiFi.<br>
```

File: proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+*3 : The USB connection is prioritized, even if both transports are connected to the same device.<br>
+*4 : If the HU recognized that current BT and new USB are connected to same device, SDL transport will switch to USB.
```
Comment (theresalech) [inline proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md:35]:
```suggestion
*4 : If the HU has recognized that current BT and new USB are connected to same device, SDL transport will switch to USB.
```

File: proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+#### 2. Addition of transport switching notification to HMI
+The following RPC, which will notify the HMI that the transport switching process is in progress, is added.
```
Comment (theresalech) [inline proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md:38]:
```suggestion
The following RPC, which will notify the HMI that the transport switching process is in progress, should be added.
```

File: proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+OEMs can implement processes related to transport switching, such as message display and screen transition, according to the HMI specifications. They can also eliminate the inconvenience for the users when using the transport switch.
+
+New RPC: `OnDeviceSwitchStatus`(HMI_API.xml)
+
+```xml
+<function name="OnDeviceSwitchStatus" messagetype="notification">
+    <description> Notify HMI of the current state of the Transport  switch </description>
```
Comment (theresalech) [inline proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md:45]:
```suggestion
    <description> Notify HMI of the current state of the transport switch </description>
```

File: proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    <param name="status" type="Common.SwitchStatus" mandatory="true"/>
+</function>
+
+<enum name="SwitchStatus">
+   <element name="Started"/>
+   <element name="Finished"/>
+</enum>
+```
+
+The below shows the new RPC (in red text) is added to the existing logic. It is necessary to add the related information regarding the new RPC, including this sequence diagram, to the "SDL Core Guide" and "HMI Documentation" on SDL Developer Portal (smartdevicelink.com).
```
Comment (theresalech) [inline proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md:55]:
```suggestion
The below shows the new RPC (in red text) added to the existing logic. It is necessary to add the related information regarding the new RPC, including this sequence diagram, to the "SDL Core Guides" and "HMI Documentation" on SDL Developer Portal (smartdevicelink.com).
```

File: proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+*1 The switching from BT to USB logic in the same device transport of iOS is already approved and adopted by SDLC ([PR](https://github.com/smartdevicelink/sdl_evolution/issues/158) and [sequence](https://cloud.githubusercontent.com/assets/11158516/26067529/3f428434-39a3-11e7-910b-6cc4069e0be3.png)).<br>
```
Comment (theresalech) [inline proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md:57]:
```suggestion
*1 The switching from BT to USB logic in the same device transport of iOS is already approved and implemented by the SDLC in SDL 0053 - Connectivity via iAP-BT and Transport Switch. See [PR](https://github.com/smartdevicelink/sdl_evolution/issues/158) and [sequence](https://cloud.githubusercontent.com/assets/11158516/26067529/3f428434-39a3-11e7-910b-6cc4069e0be3.png).<br>
```

File: proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+*2 With the current SDL standard specification, the device cannot be identified whether it is the same device or not. Therefore, device identification by the HU must be a prerequisite.
```
Comment (theresalech) [inline proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md:58]:
```suggestion
*2 With the current SDL standard specification, the device cannot identify whether it is the same device or not. Therefore, device identification by the HU must be a prerequisite.
```

File: proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+## Potential downsides
+There are no downsides to this proposal.
+
+
+## Impact on existing code
+This proposal impacts the following:
+
+- SDL Core
+- SDL Core Document
```
Comment (theresalech) [inline proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md:73]:
```suggestion
- SDL Core Documentation
```

File: proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+- HMI
+- HMI Document
```
Comment (theresalech) [inline proposals/NNNN-Clarification-of-transport-switching-rules-and-improvement-of-UX.md:75]:
```suggestion
- HMI Documentation
```

**Comment by Akihiro-Miyazaki** [comment]
Hi, @theresalech -san.
Thank you in advance for your review.
I made a commit all your suggestion.
Could you please confirm them.
Thank you.

**Pull request merged**